### PR TITLE
fix: parse value imports named type as values

### PR DESCRIPTION
**Pull request by kaicataldo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | fixes https://github.com/babel/babel/issues/11291
| Patch: Bug Fix?          | :+1:
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This PR updates the logic for parsing import types to make sure that `type` is actually a keyword rather than an identifier for a value import.

Questions:
1. I'm not sure why so many `options.json` files changes when I ran the tests with `OVERWRITE=true`. Should I revert those changes?
1. Am I missing any cases?

### CODE REVIEW

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 1]
```diff
@@ -1866,7 +1866,17 @@ export default (superClass: Class<Parser>): Class<Parser> =>
         return this.tsParseImportEqualsDeclaration(node);
       }
 
-      if (this.eatContextual("type")) {
+      const ahead = this.lookahead();
```
Comment (existentialism) [inline packages/babel-parser/src/plugins/typescript/index.js]: We should probably only do the lookahead if `this.match(tt.name)`.

Does it make sense to combine this with the above block?

```
if (this.match(tt.name)) {
  const ahead = this.lookahead();

  if (ahead.type === tt.eq) {
    return this.tsParseImportEqualsDeclaration(node);
  }

  if (
    this.state.value === "type" && 
    // import type, { a } from "b";
    ahead.type !== tt.comma &&
    // import type from "a";
    !(ahead.type === tt.name && ahead.value === "from")
  ) {
    node.importKind = "type";
  }
} else {
  node.importKind = "value";
}
```

File: packages/babel-parser/test/fixtures/estree/typescript/import-require/output.json [hunk 2]
```diff
@@ -43,6 +43,7 @@
             "column": 32
           }
         },
+        "importKind": "value",
```
Comment (existentialism) [inline packages/babel-parser/test/fixtures/estree/typescript/import-require/output.json]: I don't think we want to set `importKind` on `TSImportEqualsDeclaration`s

Comment (kaicataldo) [reply packages/babel-parser/test/fixtures/estree/typescript/import-require/output.json]: For my own understanding, why is that? Isn't it an imported value?

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 3]
```diff
@@ -1862,14 +1862,25 @@ export default (superClass: Class<Parser>): Class<Parser> =>
...
-        node.importKind = "value";
+      if (this.match(tt.name)) {
+        const ahead = this.lookahead();
+
+        if (ahead.type === tt.eq) {
+          return this.tsParseImportEqualsDeclaration(node);
+        }
+
+        if (
+          this.state.value === "type" &&
```
Comment (JLHwung) [inline packages/babel-parser/src/plugins/typescript/index.js]: should also check `!this.state.containsEsc`, which is also included in `isContexual`.
```js
import typ\u{65} typescript from "typescript";
```

Note, it seems that the TS [REPL](https://www.typescriptlang.org/play/?ts=Nightly#code/JYWwDg9gTgLgBDAnmAOgVwN4DYCsBfBZAUwGcBjKYMeAMyghDgCIkxSKqYmBuIA) does not agree with me, which should be a bug.

Comment (nicolo-ribaudo) [reply packages/babel-parser/src/plugins/typescript/index.js]: Yeah, let's assume that escapes are disallowed for now and report it to the TS team (cc @DanielRosenwasser)

File: packages/babel-parser/test/fixtures/flow/anonymous-function-types/bad_01/options.json [hunk 4]
```diff
@@ -1,3 +1,8 @@
 {
+  "sourceType": "module",
+  "plugins": [
+    "jsx",
+    "flow"
+  ],
   "throws": "Unexpected token, expected \";\" (1:29)"
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/test/fixtures/flow/anonymous-function-types/bad_01/options.json]: Could you revert all these unrelated changes to the tests? :sweat_smile:

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 5]
```diff
@@ -1862,14 +1862,25 @@ export default (superClass: Class<Parser>): Class<Parser> =>
...
+        }
+
+        if (
+          this.isContextual("type") &&
+          // import type, { a } from "b";
+          ahead.type !== tt.comma &&
+          // import type from "a";
+          !(ahead.type === tt.name && ahead.value === "from")
+        ) {
+          this.eat(tt.name);
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/plugins/typescript/index.js]: Nit: `this.next()` is enough, since we are not using the return value of `this.eat` here.

File: packages/babel-parser/test/fixtures/flow/anonymous-function-types/bad_01/options.json (hunk 4 above)
Comment (kaicataldo) [reply packages/babel-parser/test/fixtures/flow/anonymous-function-types/bad_01/options.json]: Sure thing.

File: packages/babel-plugin-transform-typescript/test/fixtures/imports/import-named-type/output.mjs [hunk 6]
```diff
@@ -0,0 +1 @@
+
```
Comment (jridgewell) [inline packages/babel-plugin-transform-typescript/test/fixtures/imports/import-named-type/output.mjs]: Shouldn't this have an output?

Comment (kaicataldo) [reply packages/babel-plugin-transform-typescript/test/fixtures/imports/import-named-type/output.mjs]: Yes - fixing now

**Comment by hzoo** [comment]
Wrote in Slack, but I think we need to check against `export type` as well? (doesn't have to be in this pr tho)

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 7]
```diff
@@ -1862,14 +1862,25 @@ export default (superClass: Class<Parser>): Class<Parser> =>
     checkDuplicateExports() {}
 
     parseImport(node: N.Node): N.AnyImport {
-      if (this.match(tt.name) && this.lookahead().type === tt.eq) {
-        return this.tsParseImportEqualsDeclaration(node);
-      }
+      if (this.match(tt.name) || this.match(tt.star) || this.match(tt.braceL)) {
```
Comment (kaicataldo) [inline packages/babel-parser/src/plugins/typescript/index.js]: I _think_ this covers all cases?

File: packages/babel-plugin-transform-typescript/test/fixtures/imports/import-named-type/output.mjs (hunk 6 above)
Comment (kaicataldo) [reply packages/babel-plugin-transform-typescript/test/fixtures/imports/import-named-type/output.mjs]: I forgot to set the `onlyRemoveTypeImports` option for these tests 🤦‍♂

File: packages/babel-parser/src/plugins/typescript/index.js (hunk 7 above)
Comment (nicolo-ribaudo) [reply packages/babel-parser/src/plugins/typescript/index.js]: Yeah it should.
